dkg-gadget/src/testing.rs

```diff
@@ -0,0 +1,1210 @@
+/// to come from the BlockChain. However, for testing purposes, we will mock
+/// the blockchain and thus send pseudo notifications to subscribing clients.
+/// To allow flexibility in the design, each [`MockClient`] will need to connect
+/// to a MockBlockchain service that periodically sends notifications to all connected
+/// clients.
+pub struct MockClient {}
+
+#[derive(Clone)]
+pub struct TestBackend {
+	inner: TestBackendState,
+}
+
+#[derive(Clone)]
+pub struct TestBackendState {
+	finality_stream: Arc<MultiSubscribableStream<FinalityNotification<TestBlock>>>,
+	import_stream: Arc<MultiSubscribableStream<ImportNotification<TestBlock>>>,
+	api: DummyApi,
+	offchain_storage: InMemOffchainStorage,
+	local_test_cases: LocalTestCases,
+}
+
+pub type LocalTestCases = Arc<Mutex<HashMap<Uuid, Option<Result<(), String>>>>>;
+
+impl TestBackend {
+	pub async fn connect<T: ToSocketAddrs>(
+		mock_bc_addr: T,
+		peer_id: PeerId,
+		api: DummyApi,
+		mut from_dkg_worker: UnboundedReceiver<(uuid::Uuid, Result<(), String>)>,
+		latest_test_uuid: Arc<RwLock<Option<Uuid>>>,
+		logger: DebugLogger,
+	) -> std::io::Result<Self> {
+		logger
+			.info(format!("0. Setting up orchestrator<=>DKG communications for peer {peer_id:?}"));
+		let socket = tokio::net::TcpStream::connect(mock_bc_addr).await?;
+		let (tx, mut rx) = dkg_mock_blockchain::transport::bind_transport::<TestBlock>(socket);
+		let tx0 = Arc::new(tokio::sync::Mutex::new(tx));
+		let tx1 = tx0.clone();
+
+		let this = TestBackend {
+			inner: TestBackendState {
+				finality_stream: Arc::new(MultiSubscribableStream::new("finality_stream")),
+				import_stream: Arc::new(MultiSubscribableStream::new("import_stream")),
+				api,
+				offchain_storage: Default::default(),
+				local_test_cases: Arc::new(Mutex::new(Default::default())),
+			},
+		};
+
+		let _this_for_dkg_listener = this.clone();
+		let logger0 = logger.clone();
+		let dkg_worker_listener = async move {
+			while let Some((trace_id, result)) = from_dkg_worker.recv().await {
+				logger0.info(format!(
+					"The client {peer_id:?} has finished test {trace_id:?}. Result: {result:?}"
+				));
+				let success = result.is_ok();
+				let packet = ProtocolPacket::ClientToBlockChain {
+					event: MockClientResponse { error: result.err(), success, trace_id },
+				};
+				tx0.lock().await.send(packet).await.unwrap();
+			}
+
+			panic!("DKG worker listener ended prematurely")
+		};
+
+		let this_for_orchestrator_rx = this.clone();
+		let orchestrator_coms = async move {
+			logger
+				.info(format!("Complete: orchestrator<=>DKG communications for peer {peer_id:?}"));
+			while let Some(packet) = rx.next().await {
+				match packet {
+					ProtocolPacket::InitialHandshake => {
+						// pong back the handshake response
+						tx1.lock()
+							.await
+							.send(ProtocolPacket::InitialHandshakeResponse { peer_id })
+							.await
+							.unwrap();
+					},
+					ProtocolPacket::BlockChainToClient { trace_id, event } => {
```

Comment:
Probably can be named `BlockchainToClient`, blockchain is one word no need for camelcasing it.